aten/src/ATen/SparseCsrTensorImpl.cpp

```diff
@@ -150,14 +150,22 @@ void SparseCsrTensorImpl::resize_as_sparse_csr_tensor_(const Tensor& src) {
       !has_symbolic_sizes_strides_,
       "resize_as_sparse_csr_tensor_ called on tensor with symbolic shape");
   set_layout(src.layout());
+  Tensor compressed_indices;
+  Tensor plain_indices;
+  std::tie(compressed_indices, plain_indices) =
+      sparse_csr::getCompressedPlainIndices(src);
+
+  // compressed indices are always crow_indices_ on the impl
   crow_indices_ = at::empty_like(
-      src.crow_indices(),
-      src.crow_indices().options(),
-      src.crow_indices().suggest_memory_format());
+      compressed_indices,
+      compressed_indices.options(),
+      compressed_indices.suggest_memory_format());
+  // plain indices are always col_indices_ on the impl
   col_indices_ = at::empty_like(
-      src.col_indices(),
-      src.col_indices().options(),
-      src.col_indices().suggest_memory_format());
+      plain_indices,
+      plain_indices.options(),
+      plain_indices.suggest_memory_format());
+  // Values are values regardless of layout
   values_ = at::empty_like(
       src.values(),
       src.values().options(),
```

Comment:
Same here. `empty_like` by default takes options of the input.

Reply:
Something that came up while writing tests:  

Resetting values like this causes the values to take on the properties of `src.values()` without modifying the properties of the `self` tensor directly. That is `self.resize_as_sparse_(other)` with self/other sparse csr will cause `self.values()` to take on dtype and device type of `other.values()` without updating the corresponding properties on `self` directly. This is before applying any of my changes here. 

- The dense analog `resize_as` will not modify `device` or `dtype` of self, only changing the shape to that of `src`. 
- The `coo` analog will throw if `device` does not match between `self` and `src`, the `dtype` of both `self` and `self.values()` remains unchanged. 
- The current `csr` only variant of this function. Will leave the `device`/`dtype` of `self` unchanged but the `device`/`dtype` of `self.values()` becomes that of `src`

It seems to me that the existing behavior here is wrong. This op should preserve the `device` and `dtype` of self, updating only the shape taken from `src`. 

Thoughts? cc @nikitaved @pearu @cpuhrsch

Reply:
> It seems to me that the existing behavior here is wrong. This op should preserve the device and dtype of self, updating only the shape taken from src.

I agree. Updating the device of crow/col_indices with src device is wrong as well.

Reply:
Or if we decide that those `dtype` and `device` be inherited from `src` We must make sure they are propagated up to `self` . You could make the argument that if we are already taking on the `layout` of `src` it makes sense to inherit the other metadata?

Reply:
For me, the name of the function defines its expected behavior.

Why do we update the layout in `resize_as`?

Reply:
In order to ensure the resized `self` is obeying the invariants of its layout. 

Suppose `self.layout` is csc and `src.layout` is csr. If `self` takes on the compressed indices of `src` they will have the shape `..., nrows + 1` If it does not also take on `csr` layout this would leave it in a state that violates the invariants of `csc` layout which should have compressed indices with shape `..., ncols +1`. That is, in the general case where the sparse dimensions have different lengths. 

Perhaps we are doing too much here all together and we should just update the shape and make nnz == 0?

Reply:
Thanks for the layout explanation, @amjames, that makes sense.

> Perhaps we are doing too much here all together and we should just update the shape and make nnz == 0?

That is a good point. Let's look at the use cases:
1. `resize_as_sparse_csr_` is used in `unary_op_out` to prepare the empty result tensor to have the input shape, this follows `copy_sparse_compressed_` call that requires the `nnz` of the input. So, making `nnz == 0` in `resize_as_sparse_csr_` would break `unary_op_out`.
2. `resize_as_sparse_csr_` is used in `add_out_sparse_csr_cuda` to prepare the output tensor size, this follows `add_out_sparse_csr` call that assumes correct `crow_indices` size and it will resize `col_indices/values`. So, this should work fine if we make `nnz==0` in `resize_as_sparse_csr_`.

Now, from a theoretical point of view, the `resize` operation is about changing the shape of an input tensor to the desired shape while reusing the existing storage (while data is discarded) --- the last point excludes setting `nnz` to zero for sparse tensors (note that we have [sparse_resize_and_clear_](https://pytorch.org/docs/master/generated/torch.Tensor.sparse_resize_and_clear_.html) for this). We have the following specifications:
- In the case of strided tensors, the shape change means setting a new shape and strides. The dimensionality of the tensor is allowed to be updated as well. (see [resize_ docs](https://pytorch.org/docs/master/generated/torch.Tensor.resize_.html))
- In the case of sparse tensors, the shape of a tensor is defined as a list of sparse and dense dimensions, plus the set of specified indices (read: `nnz`). The resize cannot change the dimensionality of the tensor (unless the input has `nnz==0`) nor the number of sparse and dense dimensions (see [sparse_resize_ docs](https://pytorch.org/docs/master/generated/torch.Tensor.sparse_resize_.html)) (unless the input has `nnz==0`), nor can specific dimensions be reduced. The restrictions for `nnz!=0` cases exist because recomputing indices is non-trivial, expensive, and usually not needed because the existing data is discarded anyway.
- In the case of sparse compressed layouts, the suggested specification is given below. The compressed nature of certain dimensions will introduce additional complications (due to crow_indices/col_indices restrictions) that need to be resolved by ensuring that tensor invariants hold and the op is computationally efficient (while discarding data is ok). This means that we shall not define `csr.resize_(new_shape)`, instead, we only define `csr.resize_as_(other_csr)` so that `other_csr` indices can be reused for the resized `csr`.

Notice that the above `resize` specifications are layout specific and one of the goals is to reuse the existing storage of the input. This means that resize should not change the layout of the input tensor.

For strided case, `input.resize_as(src)` is defined as `input.resize_(src.shape)`. So, it would make sense to reuse this definition also for sparse tensors with additional conditions on the `nnz` (to reuse the existing storage of indices/values).

In sum, taking into account what discussed above, I suggest implementing the following algorithm for `resize_as_sparse_csr_` method (using Python-like pseudocode):
```python
def resize_as_sparse_compressed_(self, other):
    assert self.layout == other.layout
    # reuse self indices storage:
    if self.crow_indices.shape != other.crow_indices.shape:
        self.crow_indices.resize_as_(other.crow_indices)
    if self.col_indices.shape != other.col_indices.shape:
        self.col_indices.resize_as_(other.col_indices)
    # optionally reuse other indices data to ensure self tensor invariants validity:
    if self.shape != other.shape or self.dense_dim() != other.dense_dim():
        self.crow_indices.copy_(other.crow_indices)
        self.col_indices.copy_(other.col_indices)
    else:
        # reusing existing self indices
        pass
    # reuse values storage:
    if self.values.shape != other.values.shape:
        self.values.resize_as_(other.values)
    self.shape = other.shape
```